### PR TITLE
[wasm][crypto] RandomNumberGenerator mapped to Web Crypto getRandomValues

### LINKED ISSUE #42727: [wasm][crypto] RandomNumberGenerator should use Web Crypto API
Uses Web Crypto API `getRandomValues` if available.

As part of issue https://github.com/dotnet/runtime/issues/40074

comment: https://github.com/dotnet/runtime/issues/40074#issuecomment-684065169

### PR DESCRIPTION
**Pull request by kjpou1**
- Uses Web Crypto API [`getRandomValues`](https://www.w3.org/TR/WebCryptoAPI/#Crypto-method-getRandomValues) if available.

- Add javascript bridge implementation library to Native source tree.  …
   - Javascript checks for crypto interface and uses `crypto.getRandomValues`
   - Add api bridge call when building for emscripten browser.
   - If we couldn't find a proper implementation, as Math.random() is not suitable we will abort.
```

ABORT: no cryptographic support found getRandomValues. Consider polyfilling it if you want to use something insecure like Math.random(), e.g. put this in a --pre-js: var crypto = { getRandomValues: function(array) { for (var i = 0; i < array.length; i++) array[i] = (Math.random()*256)|0 } };

```

closes https://github.com/dotnet/runtime/issues/42727

Part of https://github.com/dotnet/runtime/issues/40074 and https://github.com/dotnet/runtime/issues/40074#issuecomment-684065169

### CODE REVIEW

**Comment by steveisok** [comment]
Doesn't `/dev/urandom` result in a call to `Crypto.getRandomValues` via the emscripten implementation?

**Comment by kjpou1** [comment]
Looks like it does:  https://github.com/paulshapiro/emscripten/blob/master/src/library_fs.js#L1296

We can close this if we want to continue using the default implementation

File: src/libraries/Native/Unix/System.Native/pal_random.c
Comment (jkotas) [inline src/libraries/Native/Unix/System.Native/pal_random.c:59]: Can the implementation be here instead of in mono runtime?

Comment (kjpou1) [reply src/libraries/Native/Unix/System.Native/pal_random.c:59]: The code that inlines the javascript code does not generate correctly from the dotnet tree.

```
EM_JS(int32_t, mono_wasm_browser_crypto_getRandomValues, (uint8_t* buffer, int32_t bufferLength),
 {
... inline javascript code
}
```

Either way the javascript code will end up being in one of the javascript libraries that this will call as well as atjhe definition to define it. 

The 

```
int32_t mono_wasm_browser_crypto_getRandomValues(uint8_t* buffer, int32_t bufferLength);
```

Will end up being something like the following anyway:

```
EMSCRIPTEN_KEEPALIVE int32_t 
mono_wasm_browser_crypto_getRandomValues(uint8_t* buffer, int32_t bufferLength);
```

And the code actually moved into the javascript libraries.  

The implantation here is actually clearer in my opinion.  Either way let me know if one prefers the above instead.

Comment (jkotas) [reply src/libraries/Native/Unix/System.Native/pal_random.c:59]: > The code that inlines the javascript code does not generate correctly from the dotnet tree.

What is the problem? Should we make it work correctly instead?

The System.*.Native shims are meant to be standalone and not depend on the runtime. The shims have no other calls to the runtime today.

Comment (kjpou1) [reply src/libraries/Native/Unix/System.Native/pal_random.c:59]: It is the javascript code is all.  It needs to be in the javascript libraries so I will move it to there and there will be no reference to the runtime but will be linked with the javascript code implementation itself.

Comment (kjpou1) [reply src/libraries/Native/Unix/System.Native/pal_random.c:59]: Ok, referencing it that way will not work either.  Not sure how we will get it without referencing that way.

**Comment by GrabYourPitchforks** [comment]
> Falls back to `/dev/urandom` as default if the crypto library is missing.

Is this even possible? Is there any wasm host which does not provide a subtle crypto implementation? And if there is, why not force the host to polyfill as in https://github.com/dotnet/runtime/pull/42731?

**Comment by steveisok** [comment]
> > Falls back to `/dev/urandom` as default if the crypto library is missing.
> 
> Is this even possible? Is there any wasm host which does not provide a subtle crypto implementation? And if there is, why not force the host to polyfill as in #42731?

This is just about the best way to call https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues.  It's the only API that isn't promised based and we can actually invoke.  

To be honest, this already works and unless I'm missing something, I would recommend we close the PR.

**Comment by GrabYourPitchforks** [comment]
Let me rephrase the requirement:

`RandomNumberGenerator`'s built-in implementation __must__ call `Crypto.getRandomValues`. It would be better if it made this call directly, but it would also be acceptable to make this call indirectly, _assuming there's no fallback code path_. The emscripten implementation doesn't seem to fulfill this latter requirement.

File: src/libraries/Native/Unix/System.Native/pal_random.c
Comment (jkotas) [reply src/libraries/Native/Unix/System.Native/pal_random.c:59]: Both pal_random.c  and driver.c are C files. Why does the javascript work in driver.c and not in pal_random.c ?

Is it because of the build steps for each are different? If yes, can we make the build steps for pal_random.c same as the build steps for driver.c so that the javascript works in pal_random.c as well?

File: src/libraries/Native/Unix/System.Native/pal_random.c
Comment (jkotas) [inline src/libraries/Native/Unix/System.Native/pal_random.c:85]: This needs to be `#else` to meet the requirement stated by @GrabYourPitchforks

**Comment by lewing** [comment]
https://github.com/emscripten-core/emscripten/pull/12240 landed recently which will expose getentropy as a very thin wrapper over subtle crypto which is what the existing wasm /dev/random uses.  The code does fall back to a polyfill on node and fails if no polyfill is available.  I propose we stick with using /dev/random until we update emscripten then call getentropy from pal_random for wasm.  Does this satisfy all the requirements?

File: src/libraries/Native/Unix/System.Native/pal_random.c
Comment (lewing) [reply src/libraries/Native/Unix/System.Native/pal_random.c:59]: > Ok, referencing it that way will not work either. Not sure how we will get it without the reference the way it is.

What exactly are you trying and how is it failing?

**Comment by lewing** [comment]
> > Falls back to `/dev/urandom` as default if the crypto library is missing.
> 
> Is this even possible? Is there any wasm host which does not provide a subtle crypto implementation? And if there is, why not force the host to polyfill as in #42731?

This just works because emscripten falls back to node's crypro.randomBytes for /dev/random which is what the code here should do as well instead of falling back to the device.

**Review by lewing — CHANGES_REQUESTED** [review]
This has several issues, it should be implemented in terms of calling [getentropy](https://man7.org/linux/man-pages/man3/getentropy.3.html) regardless of where the implementation lives.

File: src/libraries/Native/Unix/System.Native/pal_random.c
Comment (kjpou1) [reply src/libraries/Native/Unix/System.Native/pal_random.c]: addressed in recent commits

File: src/libraries/Native/Unix/System.Native/pal_random.c
Comment (kjpou1) [reply src/libraries/Native/Unix/System.Native/pal_random.c]: This no longer inlines the JS in driver.c.

File: src/libraries/Native/Unix/System.Native/browser/pal_random_browser.h
Comment (jkotas) [inline src/libraries/Native/Unix/System.Native/browser/pal_random_browser.h:9]: This should not be `PALEXPORT`. The method is not exported from the PAL.

File: src/libraries/Native/Unix/System.Native/browser/pal_random_browser.h
Comment (jkotas) [inline src/libraries/Native/Unix/System.Native/browser/pal_random_browser.h:1]: I am not sure whether this header file is needed. I think it would be ok to have the function declaration in the .c file.

If you would like to keep it, it would be nice for it to have same name as the `.js` file.

File: src/libraries/Native/Unix/System.Native/browser/pal_entropy_bridge.js
Comment (jkotas) [inline src/libraries/Native/Unix/System.Native/browser/pal_entropy_bridge.js:1]: We do not have other OS-specific System.Native implementation in subdirectories. For example, `pal_log.m` is Apple OS specific and it does not live in subdirectory.

It would be nice to maintain same style.

File: src/libraries/Native/Unix/System.Native/browser/pal_random_browser.h
Comment (kjpou1) [reply src/libraries/Native/Unix/System.Native/browser/pal_random_browser.h]: very true.  addressed in recent commits

File: src/libraries/Native/Unix/System.Native/browser/pal_random_browser.h
Comment (kjpou1) [reply src/libraries/Native/Unix/System.Native/browser/pal_random_browser.h]: Not needed and addressed in recent commits

File: src/libraries/Native/Unix/System.Native/browser/pal_entropy_bridge.js
Comment (kjpou1) [reply src/libraries/Native/Unix/System.Native/browser/pal_entropy_bridge.js]: renamed and addressed in recent commits

File: src/libraries/Native/Unix/System.Native/pal_random.js
Comment (jkotas) [inline src/libraries/Native/Unix/System.Native/pal_random.js:17]: I do not think we should be suggesting an unsecure polyfill in the error message. 

I think we should leave it for the documentation to discuss the polyfill workarounds. The recommended workaround should be something that tries to be secure as much as possible, like the MSR library discussed in the other PR. If somebody wants to use insecure polyfill, I would leave it up to them to figure out how to do it. 

cc @GrabYourPitchforks

File: src/mono/wasm/runtime-test.js
Comment (jkotas) [inline src/mono/wasm/runtime-test.js:69]: This should have a comment that this is a simple insecure polyfill for testing purposes only.

File: src/mono/wasm/runtime-test.js
Comment (CoffeeFlux) [inline src/mono/wasm/runtime-test.js:65]: I don't think this is a functional change since typeof should always return a string. Should we be using `===` throughout our JS? I've seen a mix checked in, but if we want to default to the versions without type conversion I'll keep an eye out for that.

File: src/mono/wasm/runtime-test.js
Comment (CoffeeFlux) [inline src/mono/wasm/runtime-test.js:170]: I believe this was correct to use tabs before.

File: src/mono/wasm/runtime-test.js
Comment (CoffeeFlux) [inline src/mono/wasm/runtime-test.js:71]:
```suggestion
				buffer [i] = (Math.random () * 256) | 0;
```

**Review by CoffeeFlux — COMMENTED** [review]
LGTM modulo the things Jan mentioned. Thanks!

File: src/libraries/Native/Unix/System.Native/pal_random.c
Comment (marek-safar) [inline src/libraries/Native/Unix/System.Native/pal_random.c:70]: I think the caching is redundant

File: src/libraries/Native/Unix/System.Native/pal_random.c
Comment (marek-safar) [inline src/libraries/Native/Unix/System.Native/pal_random.c:79]: You could also just `return -1` without #else/#endif

File: src/mono/wasm/runtime-test.js
Comment (lewing) [reply src/mono/wasm/runtime-test.js:65]: We should clean up these checks in general the match the js version we require but this is fine.

**Review by lewing — COMMENTED** [review]
Address the open reviews and this looks good.

File: src/mono/wasm/runtime-test.js
Comment (kjpou1) [reply src/mono/wasm/runtime-test.js:70]: Addressed

File: src/mono/wasm/runtime-test.js
Comment (kjpou1) [reply src/mono/wasm/runtime-test.js]: addressed

File: src/libraries/Native/Unix/System.Native/pal_random.c
Comment (kjpou1) [reply src/libraries/Native/Unix/System.Native/pal_random.c:70]: It follows what is already in place and keeps us from calling external routine again unless I am missing something.  When we were aborting from the js code I would agree but recent mod now returns the -1 from the routine which will allow the managed exception be called.

Let me know if this needs to changed.

File: src/libraries/Native/Unix/System.Native/pal_random.c
Comment (kjpou1) [reply src/libraries/Native/Unix/System.Native/pal_random.c:79]: #else was recommended above in previous comment.

File: src/libraries/Native/Unix/System.Native/pal_random.js
Comment (kjpou1) [reply src/libraries/Native/Unix/System.Native/pal_random.js:17]: Changed to let normal managed exception to handle error.

```
WASM-ERR: 
WASM-ERR: Unhandled Exception:
WASM-ERR: System.Security.Cryptography.CryptographicException: Error occurred during a cryptographic operation.
WASM-ERR:    at Interop.GetCryptographicallySecureRandomBytes(Byte* buffer, Int32 length)
WASM-ERR:    at System.Security.Cryptography.RandomNumberGeneratorImplementation.GetBytes(Byte* pbBuffer, Int32 count)
WASM-ERR:    at System.Security.Cryptography.RandomNumberGeneratorImplementation.GetBytes(Span`1 data)
WASM-ERR:    at System.Security.Cryptography.RandomNumberGeneratorImplementation.GetNonZeroBytes(Span`1 data)
WASM-ERR:    at System.Security.Cryptography.RandomNumberGeneratorImplementation.GetNonZeroBytes(Byte[] data)

```

Does this work?

**Review by jkotas — APPROVED** [review]

**Review by lewing — APPROVED** [review]

**Pull request merged**
